[Link] Adds internal authentication only mode to `LinkController`

carlosmuvi-stripe opened the pull request:
# Summary
- Adds `Authentication` Launch mode. This halts the Link flow as soon as the user authenticates / signs up. 
- Adds new `presentForAuthentication` entry point (goal is to use it just for onRamp)

[full_controller.webm](https://github.com/user-attachments/assets/259588e2-7b28-4c15-84c2-b7b4ea559c8f)

# Motivation
Onramp support for verification only flows.

# Testing
- [x] Modified tests
- [x] Manually verified

carlosmuvi-stripe commented on paymentsheet-example/src/main/java/com/stripe/android/paymentsheet/example/playground/LinkControllerUi.kt:
```diff
@@ -156,9 +162,24 @@ internal fun LinkControllerUi(
     }
 }
 
+@Composable
+private fun LinkControllerPlaygroundState.linkControllerError(): Throwable? = listOf(
+    (presentPaymentMethodsResult as? LinkController.PresentPaymentMethodsResult.Failed)?.error,
+    (lookupConsumerResult as? LinkController.LookupConsumerResult.Failed)?.error,
+    (createPaymentMethodResult as? LinkController.CreatePaymentMethodResult.Failed)?.error,
+    (presentForAuthenticationResult as? LinkController.PresentForAuthenticationResult.Failed)?.error,
+).firstOrNull { it != null }
+
```
no logic changes here, just extracted error resolution to a method.

carlosmuvi-stripe commented:
@Twigz tagging you just for visibility : )

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/signup/SignUpViewModel.kt:
```diff
@@ -278,7 +294,8 @@ internal class SignUpViewModel @Inject constructor(
         fun factory(
             parentComponent: NativeLinkComponent,
             navigateAndClearStack: (LinkScreen) -> Unit,
-            moveToWeb: () -> Unit
+            moveToWeb: () -> Unit,
+            dismissWithResult: (LinkActivityResult) -> Unit
```
It seems to me that passing in a lambda to a VM factory is an anti-pattern since it can easily leak the Activity (lambda captures an Activity; ViewModel is reused on configuration change). Fortunately, the lambdas we're passing always happen to capture only other Activity ViewModels, but I think it's a footgun we should disarm.

carlosmuvi-stripe replied:
100% agreed. We recently refactored navigation away from this pattern in Link in favor of `NavigationManager` (flow-based) - I'll move all existing lambdas (unfortunately that's the current pattern) to a similar approach in a separate PR if that works!

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/LinkControllerViewModel.kt:
```diff
@@ -91,7 +99,117 @@ internal class LinkControllerViewModel @Inject constructor(
+        }
+    }
+
+    internal fun onPresentForAuthentication(
```
add unit tests for this?

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/LinkController.kt:
```diff
@@ -60,6 +60,27 @@ class LinkController @Inject internal constructor(
+     * matches an existing Link account, the user will be able to authenticate with that account.
+     * If null, the user will need to sign in or create a Link account.
+     */
+    fun presentForAuthentication(email: String?) {
```
I know the class is already restricted, but out of abundance of caution can you restrict this function specifically since we intend for it to be internal for the foreseeable future?
```suggestion
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    fun presentForAuthentication(email: String?) {
```

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/LinkControllerViewModel.kt:
```diff
@@ -91,7 +99,117 @@ internal class LinkControllerViewModel @Inject constructor(
+        val accountUpdate = linkAccountHolder.linkAccountInfo.value
+            .takeIf { email == state.presentedForEmail || email == it.account?.email }
+            ?: LinkAccountUpdate.Value(null)
+        if (accountUpdate.account == null) {
+            linkAccountHolder.set(accountUpdate)
+        }
```
Can we add a test for this?

lng-stripe reviewed (changes requested):
Looking pretty good! A few things:
 1. Can you update the demo in the PR description? I think you may have uploaded the wrong one.
 2. Could you add/update tests for the ViewModels?
 3. I have some FUD around passing lambdas to a VM factory, but since it's an established pattern maybe we should address it in follow-ups.

carlosmuvi-stripe commented:
@lng-stripe addressed feeback!

- Added tests
- Added video authenticating only, then choosing a payment method showcasing being already logged in
- Replied re the callback approach - that's on the radar, I'll tackle it separately if that works
